flink-ml-lib/src/main/java/org/apache/flink/ml/common/util/QuantileSummary.java

```diff
@@ -0,0 +1,400 @@
+     * need to call compress() before using it.
+     *
+     * @param percentile The target percentile.
+     * @return The corresponding approximate quantile.
+     */
+    public double query(double percentile) {
+        return query(new double[] {percentile})[0];
+    }
+
+    /**
+     * Runs a query for a given sequence of percentiles. The query can only be run on a compressed
+     * summary, you need to call compress() before using it.
+     *
+     * @param percentiles A list of the target percentiles.
+     * @return A list of the corresponding approximate quantiles, in the same order as the input.
+     */
+    public double[] query(double[] percentiles) {
+        Arrays.stream(percentiles)
+                .forEach(
+                        x ->
+                                Preconditions.checkState(
+                                        x >= 0 && x <= 1.0,
+                                        "percentile should be in the range [0.0, 1.0]."));
+        Preconditions.checkState(
+                headBuffer.isEmpty(),
+                "Cannot operate on an uncompressed summary, call compress() first.");
+        Preconditions.checkState(
+                sampled != null && !sampled.isEmpty(),
+                "Cannot query percentiles without any records inserted.");
+        double targetError = Long.MIN_VALUE;
+        for (StatsTuple tuple : sampled) {
+            targetError = Math.max(targetError, (tuple.delta + tuple.g));
+        }
+        targetError = targetError / 2;
+        Map<Double, Integer> zipWithIndex = new HashMap<>(percentiles.length);
+        IntStream.range(0, percentiles.length).forEach(i -> zipWithIndex.put(percentiles[i], i));
+
+        int index = 0;
+        long minRank = sampled.get(0).g;
+        double[] sorted = Arrays.stream(percentiles).sorted().toArray();
+        double[] result = new double[percentiles.length];
+
+        for (int i = 0; i < sorted.length; i++) {
+            int percentileIndex = zipWithIndex.get(sorted[i]);
+            if (sorted[i] <= relativeError) {
+                result[percentileIndex] = sampled.get(0).value;
+            } else if (sorted[i] >= 1 - relativeError) {
+                result[percentileIndex] = sampled.get(sampled.size() - 1).value;
+            } else {
+                QueryResult queryResult =
+                        findApproximateQuantile(index, minRank, targetError, sorted[i]);
+                index = queryResult.index;
+                minRank = queryResult.minRankAtIndex;
+                result[percentileIndex] = queryResult.percentile;
+            }
+        }
+        return result;
+    }
+
+    /**
+     * Check whether the QuantileSummary has inserted rows. Running query on an empty
+     * QuantileSummary would cause {@link java.lang.IllegalStateException}.
+     *
+     * @return True if the QuantileSummary is empty, otherwise false.
+     */
+    public boolean isEmpty() {
```

Comment:
Should it be `private`?

Reply:
It's a public API. Running query on an empty QuantileSummary is illegal and exceptions would be thrown. So we'd better provide a public API to check if the QuantileSummary is empty.

Reply:
@jiangxin369 Since this is public API, should we add a test so that every public API's behavior can be covered by unit tests?

Reply:
Sure

Reply:
Is this comment addressed? QuantileSummaryTest does not seem to have any test that calls `IsEmpty()`, right?

Reply:
Thanks for reminding me, I lost it.